### ISSUE TITLE
ADC INL compensation (probably) wrong

### ISSUE THREAD

**Issue by mordae** [Len42/Dexy#5]
Hi, I don't think this is correct:

https://github.com/Len42/Dexy/blob/bb95044699c291ebf2c83961f4cac21d869a0346/firmware/AdcInput.cpp#L17

The datasheet "sawtooth" plot details integral non-linearity. That is an absolute error measure. Your code adds the absolute errors together, resulting in a nonsensical value. I suggest creating a table of 256 values long segments along with their absolute errors and subtracting those instead.

Something along the lines of:

```c
static const int errors[16] = { 0, -2, 6, 4, ... };

int raw_value = adc_read();
int calib_value = raw_value - errors[raw_value >> 8];
```

It is possible that I am mistaken, though. If you have measured the ADC against a voltage source and found your code works well, please let me know.

**Comment by Len42** [comment]
I did calibrate the tuning using a keyboard and a tuner (and I verified that the CV from the keyboard is accurate). The code as written does get rid of the linearity errors I was seeing. The tuning isn't perfect but the remaining inaccuracy seems to be due to quantization and analog noise.

Here's how I came up with that code: I noticed that as I played notes up the keyboard, there were sudden jumps in the ADC value (and the pitch) between certain notes. I found that those jumps corresponded to the jumps in the INL graph. By making the adjustments in linearizeAdcInput() at those specific ADC values, I was able to get a linear CV-to-ADC response, so I could fit a linear ADC-to-pitch conversion that gives overall linear CV-to-pitch response.

Really, a better ADC is needed, with at least 14 bits of precision and better linearity.
